### CS3405 Homework 4 DFT and FFT
#### 梁浩祥 106034045
##### 1. Implement your own DFT, IDFT, FFT, IFFT
##### 2. Extract three data sequences, N=1024, N=4096, N=2**15 from each of violin22k, gtr-nylon22 and pno-cs24k. That is, you have nine input data. Call them v1k, v4k, v32k, g1k, g4k, g32k, p1k, p4k, p32k
#### 3. Test your programs by running each of nine sequences thru DFT-->IDFT, DFT-->IFFT, FFT-->IFFT, FFT-->IDFT
#### 4. Make sure all are correct and compare the run times.

In [11]:
%matplotlib inline
import matplotlib.pyplot as plt
import pylab
import numpy as np
import math
import cmath
import scipy.io.wavfile as sw
import time

In [79]:
def dft(x):
    N=len(x)
    X=np.array([sum(x*np.array([cmath.exp(-1j*2*cmath.pi*m/N*n) for n in range(N)])) for m in range(N)])
    return X
def idft(x):
    N=len(x)
    X=np.array([sum(x*np.array([cmath.exp(1j*2*cmath.pi*m/N*n) for m in range(N)]))/N for n in range(N)])
    return X
def fft(x):
    N = len(x)
    Half_N = N // 2
    X = np.zeros(N, dtype = complex)
    if N == 1:
        X[0] = x[0]
    else:
        X_even = fft(x[0::2]) * Half_N
        X_odd = fft(x[1::2]) * Half_N
        for m in range(N):
            X[m] = X_even[m%Half_N] + X_odd[m%Half_N] * cmath.exp(-1j*2*math.pi*m/N)
    return X / N
        
def __ifft(X):
    N = len(X)
    Half_N = N // 2
    x = np.zeros(N, dtype = complex)
    if N == 1:
        x[0] = X[0]
    else:
        x_even = __ifft(X[0::2])
        x_odd = __ifft(X[1::2])
        for n in range(N):
            x[n] = x_even[n%Half_N] + x_odd[n%Half_N] * cmath.exp(1j*2*math.pi*n/N)
    return x
def ifft(X):
    return __ifft(X) / len(X)
def verify(a,b,N):
    for i in range(1,N):
        if (not (((a[2048+i]-b[i].real)**2)**0.5)<0.1):
            return False
    return True

Implement violin22k file using n=1024

In [97]:
fs, v1k = sw.read('violin22k.wav')
t=time.time()
x = dft(v1k[2048:2048+1024])
xre = idft(x)
t2=time.time()
if (verify(v1k,xre,1024)):
    print ("Cost time      (v1k)(DFT-->IDFT):" ,t2-t)
else:
    print ("Error : reconstruct error")
t=time.time()
x = dft(v1k[2048:2048+1024])
xre = ifft(x)
t2=time.time()
if (verify(v1k,xre,1024)):
    print ("Cost time      (v1k)(DFT-->IFFT):" ,t2-t)
else:
    print ("Error : reconstruct error")
t=time.time()
x = fft(v1k[2048:2048+1024])
xre = __ifft(x)
t2=time.time()
if (verify(v1k,xre,1024)):
    print ("Cost time      (v1k)(FFT-->IFFT):" ,t2-t)
else:
    print ("Error : reconstruct error")
t=time.time()
x = fft(v1k[2048:2048+1024]) * len(x)
xre = idft(x)
t2=time.time()
if (verify(v1k,xre,1024)):
    print ("Cost time      (v1k)(FFT-->IDFT):" ,t2-t)
else:
    print ("Error : reconstruct error")


Cost time      (v1k)(DFT-->IDFT): 1.382645606994629
Cost time      (v1k)(DFT-->IFFT): 0.6850447654724121
Cost time      (v1k)(FFT-->IFFT): 0.035916805267333984
Cost time      (v1k)(FFT-->IDFT): 0.7060012817382812


Implement violin22k file using n=4096

In [95]:
fs, v4k = sw.read('violin22k.wav')
R=4096
t=time.time()
x = dft(v4k[2048:2048+R])
xre = idft(x)
t2=time.time()
if (verify(v1k,xre,R)):
    print ("Cost time      (v4k)(DFT-->IDFT):" ,t2-t)
else:
    print ("Error : reconstruct error")

t=time.time()
x = dft(v4k[2048:2048+R])
xre = ifft(x)
t2=time.time()
if (verify(v1k,xre,R)):
    print ("Cost time      (v4k)(DFT-->IFFT):" ,t2-t)
else:
    print ("Error : reconstruct error")

t=time.time()
x = fft(v4k[2048:2048+R])
xre = __ifft(x)
t2=time.time()
if (verify(v1k,xre,R)):
    print ("Cost time      (v4k)(FFT-->IFFT):" ,t2-t)
else:
    print ("Error : reconstruct error")
    
t=time.time()
x = fft(v4k[2048:2048+R]) * len(x)
xre = idft(x)
t2=time.time()
if (verify(v1k,xre,R)):
    print ("Cost time      (v4k)(FFT-->IDFT):" ,t2-t)
else:
    print ("Error : reconstruct error")

Cost time      (v4k)(DFT-->IDFT): 19.35226035118103
Cost time      (v4k)(DFT-->IFFT): 9.764917850494385
Cost time      (v4k)(FFT-->IFFT): 0.13563776016235352
Cost time      (v4k)(FFT-->IDFT): 10.061075448989868


Implement violin22k file using n=2**15

In [96]:
fs, v32k = sw.read('violin22k.wav')
R=2**15
t=time.time()
x = dft(v32k[2048:2048+R])
xre = idft(x)
t2=time.time()
if (verify(v1k,xre,R)):
    print ("Cost time      (v32k)(DFT-->IDFT):" ,t2-t)
else:
    print ("Error : reconstruct error")

t=time.time()
x = dft(v32k[2048:2048+R])
xre = ifft(x)
t2=time.time()
if (verify(v1k,xre,R)):
    print ("Cost time      (v32k)(DFT-->IFFT):" ,t2-t)
else:
    print ("Error : reconstruct error")

t=time.time()
x = fft(v32k[2048:2048+R])
xre = __ifft(x)
t2=time.time()
if (verify(v1k,xre,R)):
    print ("Cost time      (v32k)(FFT-->IFFT):" ,t2-t)
else:
    print ("Error : reconstruct error")
    
t=time.time()
x = fft(v32k[2048:2048+R]) * len(x)
xre = idft(x)
t2=time.time()
if (verify(v1k,xre,R)):
    print ("Cost time      (v32k)(FFT-->IDFT):" ,t2-t)
else:
    print ("Error : reconstruct error")

Cost time      (v32k)(DFT-->IDFT): 1246.9222702980042
Cost time      (v32k)(DFT-->IFFT): 617.5702366828918
Cost time      (v32k)(FFT-->IFFT): 1.2725977897644043
Cost time      (v32k)(FFT-->IDFT): 630.991977930069


Implement gtr-nylon22 file using n=1024

In [89]:
fs, g1k = sw.read('gtr-nylon22.wav')
R=1024
t=time.time()
x = dft(g1k[2048:2048+R])
xre = idft(x)
t2=time.time()
if (verify(g1k,xre,R)):
    print ("Cost time      (g1k)(DFT-->IDFT):" ,t2-t)
else:
    print ("Error : reconstruct error")

t=time.time()
x = dft(g1k[2048:2048+R])
xre = ifft(x)
t2=time.time()
if (verify(g1k,xre,R)):
    print ("Cost time      (g1k)(DFT-->IFFT):" ,t2-t)
else:
    print ("Error : reconstruct error")

t=time.time()
x = fft(g1k[2048:2048+R])
xre = __ifft(x)
t2=time.time()
if (verify(g1k,xre,R)):
    print ("Cost time      (g1k)(FFT-->IFFT):" ,t2-t)
else:
    print ("Error : reconstruct error")
    
t=time.time()
x = fft(g1k[2048:2048+R]) * len(x)
xre = idft(x)
t2=time.time()
if (verify(g1k,xre,R)):
    print ("Cost time      (g1k)(FFT-->IDFT):" ,t2-t)
else:
    print ("Error : reconstruct error")

Cost time      (g1k)(DFT-->IDFT): 1.4560306072235107
Cost time      (g1k)(DFT-->IFFT): 0.7640340328216553
Cost time      (g1k)(FFT-->IFFT): 0.0359957218170166
Cost time      (g1k)(FFT-->IDFT): 0.7799732685089111


Implement gtr-nylon22 file using n=4096

In [90]:
fs, g4k = sw.read('gtr-nylon22.wav')
R= 4096
t=time.time()
x = dft(g4k[2048:2048+R])
xre = idft(x)
t2=time.time()
if (verify(g4k,xre,R)):
    print ("Cost time      (g4k)(DFT-->IDFT):" ,t2-t)
else:
    print ("Error : reconstruct error")

t=time.time()
x = dft(g4k[2048:2048+R])
xre = ifft(x)
t2=time.time()
if (verify(g4k,xre,R)):
    print ("Cost time      (g4k)(DFT-->IFFT):" ,t2-t)
else:
    print ("Error : reconstruct error")

t=time.time()
x = fft(g4k[2048:2048+R])
xre = __ifft(x)
t2=time.time()
if (verify(g4k,xre,R)):
    print ("Cost time      (g4k)(FFT-->IFFT):" ,t2-t)
else:
    print ("Error : reconstruct error")
    
t=time.time()
x = fft(g4k[2048:2048+R]) * len(x)
xre = idft(x)
t2=time.time()
if (verify(g4k,xre,R)):
    print ("Cost time      (g4k)(FFT-->IDFT):" ,t2-t)
else:
    print ("Error : reconstruct error")

Cost time      (g4k)(DFT-->IDFT): 26.776000261306763
Cost time      (g4k)(DFT-->IFFT): 12.220002889633179
Cost time      (g4k)(FFT-->IFFT): 0.17199134826660156
Cost time      (g4k)(FFT-->IDFT): 12.804014921188354


Implement gtr-nylon22 file using n=2**15

In [91]:
fs, g32k = sw.read('gtr-nylon22.wav')
R= 2**15
t=time.time()
x = dft(g32k[2048:2048+R])
xre = idft(x)
t2=time.time()
if (verify(g32k,xre,R)):
    print ("Cost time      (g32k)(DFT-->IDFT):" ,t2-t)
else:
    print ("Error : reconstruct error")

t=time.time()
x = dft(g32k[2048:2048+R])
xre = ifft(x)
t2=time.time()
if (verify(g32k,xre,R)):
    print ("Cost time      (g32k)(DFT-->IFFT):" ,t2-t)
else:
    print ("Error : reconstruct error")

t=time.time()
x = fft(g32k[2048:2048+R])
xre = __ifft(x)
t2=time.time()
if (verify(g32k,xre,R)):
    print ("Cost time      (g32k)(FFT-->IFFT):" ,t2-t)
else:
    print ("Error : reconstruct error")
    
t=time.time()
x = fft(g32k[2048:2048+R]) * len(x)
xre = idft(x)
t2=time.time()
if (verify(g32k,xre,R)):
    print ("Cost time      (g32k)(FFT-->IDFT):" ,t2-t)
else:
    print ("Error : reconstruct error")

Cost time      (g32k)(DFT-->IDFT): 1426.8474249839783
Cost time      (g32k)(DFT-->IFFT): 825.7731926441193
Cost time      (g32k)(FFT-->IFFT): 1.9839985370635986
Cost time      (g32k)(FFT-->IDFT): 803.7169151306152


Implement pno-cs24k file using n=1024

In [92]:
fs, p1k = sw.read('pno-cs24k.wav')
R= 1024
t=time.time()
x = dft(p1k[2048:2048+R])
xre = idft(x)
t2=time.time()
if (verify(p1k,xre,R)):
    print ("Cost time      (p1k)(DFT-->IDFT):" ,t2-t)
else:
    print ("Error : reconstruct error")

t=time.time()
x = dft(p1k[2048:2048+R])
xre = ifft(x)
t2=time.time()
if (verify(p1k,xre,R)):
    print ("Cost time      (p1k)(DFT-->IFFT):" ,t2-t)
else:
    print ("Error : reconstruct error")

t=time.time()
x = fft(p1k[2048:2048+R])
xre = __ifft(x)
t2=time.time()
if (verify(p1k,xre,R)):
    print ("Cost time      (p1k)(FFT-->IFFT):" ,t2-t)
else:
    print ("Error : reconstruct error")
    
t=time.time()
x = fft(p1k[2048:2048+R]) * len(x)
xre = idft(x)
t2=time.time()
if (verify(p1k,xre,R)):
    print ("Cost time      (p1k)(FFT-->IDFT):" ,t2-t)
else:
    print ("Error : reconstruct error")

Cost time      (p1k)(DFT-->IDFT): 1.4240002632141113
Cost time      (p1k)(DFT-->IFFT): 0.8199999332427979
Cost time      (p1k)(FFT-->IFFT): 0.04002737998962402
Cost time      (p1k)(FFT-->IDFT): 0.755974292755127


Implement pno-cs24k file using n=4096

In [93]:
fs, p4k = sw.read('pno-cs24k.wav')
R= 4096
t=time.time()
x = dft(p4k[2048:2048+R])
xre = idft(x)
t2=time.time()
if (verify(p4k,xre,R)):
    print ("Cost time      (p4k)(DFT-->IDFT):" ,t2-t)
else:
    print ("Error : reconstruct error")

t=time.time()
x = dft(p4k[2048:2048+R])
xre = ifft(x)
t2=time.time()
if (verify(p4k,xre,R)):
    print ("Cost time      (p4k)(DFT-->IFFT):" ,t2-t)
else:
    print ("Error : reconstruct error")

t=time.time()
x = fft(p4k[2048:2048+R])
xre = __ifft(x)
t2=time.time()
if (verify(p4k,xre,R)):
    print ("Cost time      (p4k)(FFT-->IFFT):" ,t2-t)
else:
    print ("Error : reconstruct error")
    
t=time.time()
x = fft(p4k[2048:2048+R]) * len(x)
xre = idft(x)
t2=time.time()
if (verify(p4k,xre,R)):
    print ("Cost time      (p4k)(FFT-->IDFT):" ,t2-t)
else:
    print ("Error : reconstruct error")

Cost time      (p4k)(DFT-->IDFT): 26.07200574874878
Cost time      (p4k)(DFT-->IFFT): 11.684001922607422
Cost time      (p4k)(FFT-->IFFT): 0.15999889373779297
Cost time      (p4k)(FFT-->IDFT): 12.119971513748169


Implement pno-cs24k file using n=2**15

In [94]:
fs, p32k = sw.read('pno-cs24k.wav')
R= 2**15
t=time.time()
x = dft(p32k[2048:2048+R])
xre = idft(x)
t2=time.time()
if (verify(p32k,xre,R)):
    print ("Cost time      (p32k)(DFT-->IDFT):" ,t2-t)
else:
    print ("Error : reconstruct error")

t=time.time()
x = dft(p32k[2048:2048+R])
xre = ifft(x)
t2=time.time()
if (verify(p32k,xre,R)):
    print ("Cost time      (p32k)(DFT-->IFFT):" ,t2-t)
else:
    print ("Error : reconstruct error")

t=time.time()
x = fft(p32k[2048:2048+R])
xre = __ifft(x)
t2=time.time()
if (verify(p32k,xre,R)):
    print ("Cost time      (p32k)(FFT-->IFFT):" ,t2-t)
else:
    print ("Error : reconstruct error")
    
t=time.time()
x = fft(p32k[2048:2048+R]) * len(x)
xre = idft(x)
t2=time.time()
if (verify(p32k,xre,R)):
    print ("Cost time      (p32k)(FFT-->IDFT):" ,t2-t)
else:
    print ("Error : reconstruct error")

Cost time      (p32k)(DFT-->IDFT): 1334.5496907234192
Cost time      (p32k)(DFT-->IFFT): 623.0711734294891
Cost time      (p32k)(FFT-->IFFT): 1.3035156726837158
Cost time      (p32k)(FFT-->IDFT): 632.6290090084076


# 比較：
* DFT-->IDFT
    * Cost time      (p32k)(DFT-->IDFT): 1334.5496907234192
    * Cost time      (p4k)(DFT-->IDFT): 26.07200574874878
    * Cost time      (p1k)(DFT-->IDFT): 1.4240002632141113
    * Cost time      (g32k)(DFT-->IDFT): 1426.8474249839783
    * Cost time      (g4k)(DFT-->IDFT): 26.776000261306763
    * Cost time      (g1k)(DFT-->IDFT): 1.4560306072235107
    * Cost time      (v32k)(DFT-->IDFT): 1246.9222702980042
    * Cost time      (v4k)(DFT-->IDFT): 19.35226035118103
    * Cost time      (v1k)(DFT-->IDFT): 1.382645606994629
    
* DFT-->IFFT
    * Cost time      (p32k)(DFT-->IFFT): 623.0711734294891
    * Cost time      (p4k)(DFT-->IFFT): 11.684001922607422
    * Cost time      (p1k)(DFT-->IFFT): 0.8199999332427979
    * Cost time      (g32k)(DFT-->IFFT): 825.7731926441193
    * Cost time      (g4k)(DFT-->IFFT): 12.220002889633179
    * Cost time      (g1k)(DFT-->IFFT): 0.7640340328216553
    * Cost time      (v32k)(DFT-->IFFT): 617.5702366828918
    * Cost time      (v4k)(DFT-->IFFT): 9.764917850494385
    * Cost time      (v1k)(DFT-->IFFT): 0.6850447654724121
    
* FFT-->IFFT
    * Cost time      (p32k)(FFT-->IFFT): 1.3035156726837158
    * Cost time      (p4k)(FFT-->IFFT): 0.15999889373779297
    * Cost time      (p1k)(FFT-->IFFT): 0.04002737998962402
    * Cost time      (g32k)(FFT-->IFFT): 1.9839985370635986
    * Cost time      (g4k)(FFT-->IFFT): 0.17199134826660156
    * Cost time      (g1k)(FFT-->IFFT): 0.0359957218170166
    * Cost time      (v32k)(FFT-->IFFT): 1.2725977897644043
    * Cost time      (v4k)(FFT-->IFFT): 0.13563776016235352
    * Cost time      (v1k)(FFT-->IFFT): 0.035916805267333984
    
* FFT-->IDFT
    * Cost time      (p32k)(FFT-->IDFT): 632.6290090084076
    * Cost time      (p4k)(FFT-->IDFT): 12.119971513748169
    * Cost time      (p1k)(FFT-->IDFT): 0.755974292755127
    * Cost time      (g32k)(FFT-->IDFT): 803.7169151306152
    * Cost time      (g4k)(FFT-->IDFT): 12.804014921188354    
    * Cost time      (g1k)(FFT-->IDFT): 0.7799732685089111
    * Cost time      (v32k)(FFT-->IDFT): 630.991977930069
    * Cost time      (v4k)(FFT-->IDFT): 10.061075448989868
    * Cost time      (v1k)(FFT-->IDFT): 0.7060012817382812
    
由此可以看出檔案的不同對時間並沒有顯著的影響，而時間的消耗為 DFT-->IDFT > DFT-->IFFT = FFT-->IDFT > FFT-->IFFT，
且FFT在執行時間上有顯著的減少。難怪FFT可以成為十大重要的演算法。